contracts/hydro/src/contract.rs

```diff
@@ -2759,6 +2778,154 @@ pub fn convert_lockup_to_dtoken_reply(
+        }
+
+        let coin_as_base = Decimal::from_ratio(coin.amount, Uint128::one()).checked_mul(ratio)?;
+
+        available_base_token = available_base_token.checked_add(coin_as_base.to_uint_floor())?;
+
+        fund_conversion.push((coin.denom.clone(), coin.amount, ratio));
+    }
+
+    // Step 4: Determine how much to deduct and update state
+    // Check if payment is full or partial
+    if available_base_token >= slash_base_token {
+        LOCKS_PENDING_SLASHES.remove(deps.storage, lockup.lock_id);
+    } else {
+        let remaining_base = slash_base_token.checked_sub(available_base_token)?;
+
+        let remaining_original = Decimal::from_ratio(remaining_base, Uint128::one())
+            .checked_div(slash_ratio)?
+            .to_uint_floor();
+        LOCKS_PENDING_SLASHES.save(deps.storage, lockup.lock_id, &remaining_original)?;
+    }
+
+    // Step 5: Figure out what was used and what should be refunded
+    let mut funds_used: HashMap<String, Uint128> = HashMap::new();
+    let mut remaining_base_token = slash_base_token.min(available_base_token);
+
+    for (denom, original_amount, ratio) in &fund_conversion {
+        if remaining_base_token.is_zero() {
+            break;
+        }
+        let coin_as_base =
+            Decimal::from_ratio(*original_amount, Uint128::one()).checked_mul(*ratio)?;
+        let take_base = coin_as_base.to_uint_floor().min(remaining_base_token);
+        let take_original = Decimal::from_ratio(take_base, Uint128::one()).checked_div(*ratio)?;
+
+        funds_used.insert(denom.clone(), take_original.to_uint_floor());
+        remaining_base_token = remaining_base_token.checked_sub(take_base)?;
+    }
+
+    // Step 6: Prepare forward and refund messages
+    let mut bank_msgs: Vec<BankMsg> = vec![];
+    for (denom, original_amount, _) in &fund_conversion {
+        let used_amount = funds_used.get(denom).copied().unwrap_or(Uint128::zero());
+        let sent_amount = *original_amount;
+        if !used_amount.is_zero() {
+            bank_msgs.push(BankMsg::Send {
+                to_address: constants.slash_tokens_receiver_addr.clone(),
+                amount: vec![Coin {
+                    denom: denom.clone(),
+                    amount: used_amount,
+                }],
+            });
+        }
+        if sent_amount > used_amount {
+            let refund_amount = sent_amount.checked_sub(used_amount)?;
+            bank_msgs.push(BankMsg::Send {
+                to_address: info.sender.to_string(),
+                amount: vec![Coin {
+                    denom: denom.clone(),
+                    amount: refund_amount,
+                }],
+            });
+        }
+    }
+
+    let coins_spent_str = funds_used
+        .iter()
+        .map(|(denom, amount)| format!("{amount}{denom}"))
+        .collect::<Vec<String>>()
+        .join(",");
+
+    Ok(Response::new()
+        .add_attribute("action", "buyout_pending_slash")
+        .add_attribute("sender", info.sender)
```

Comment:
needs more info in the Response here. I would add "amount left to repay" as an attribute here